Fix 92551 with point touching edge

craigtaverner opened the pull request:
If an H3 cell touches an outside polygon edge, it should be considered intersecting.
(labeled `non-issue` because this is a fix to an unreleased feature)

Fixes #92551

craigtaverner committed:
Fix 92551 with point touching edge

If an H3 cell touches an outside polygon edge, it should be considered
intersecting.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexVisitor.java:
```diff
@@ -297,21 +297,21 @@ private GeoRelation relateTriangle(
 
         boolean within = false;
-        if (edgeIntersectsQuery(aX, aY, bX, bY, false)) {
+        if (edgeIntersectsQuery(aX, aY, bX, bY, ab)) {
```
From other implementation, I am more tempted to set the flag to be always true. I am afraid of edge cases if all edges are false.

craigtaverner committed:
Always include border for point crossing line

When a triangle tests if a point intersects a line, the case of the
point just touching the line also counts as intersecting.

craigtaverner committed:
Fix issue of long rectangles contained in h3 cells

A long rectangle can cross the cell edges due to distortion,
so we need to check for this case. Previously we only considered
this for dateline crossings.

iverase reviewed (approved):
LGTM

nice that we found those bugs early in the process :)

craigtaverner committed:
Extended fix to more cases

However, it turned out that the issue was already fixed in another
PR, so this commit merely improves test coverage to the known
tricky cases.

Also the calculation of expected results in GeoHexTilerTests was
optimized to not repeatedly create the same tilers.
